### PR TITLE
[RFT]hostapd: add mbedtls variant

### PR DESCRIPTION
**Pull request by neheb**
@gstrauss has been kind enough to write an mbedTLS implementation for us to use.

Testing is mostly what's needed. If successful, the plan is to default to it instead of wolfSSL which seems to have problems with ABI stability.

### CODE REVIEW

**Comment by ynezz** [comment]
@neheb @gstrauss thanks a lot for this! 

> now contains a mostly functional hostapd and wpa_suppplicant using mbedtls, [Edit: not yet] including SAE and DPP.
> plan to post to the hostapd mailing list early next week, if not sooner.

BTW what's the plan/status WRT above?

**Comment by gstrauss** [comment]
As noted in https://github.com/openwrt/openwrt/issues/10303#issuecomment-1247311546:
Posted to hostap mailing list: http://lists.infradead.org/pipermail/hostap/2022-September/040794.html

The patches currently contain support for SAE, though testing is needed.  More details are in the link sent to the hostap mailing list.

**Comment by neheb** [comment]
Initial results are not good:

```
Thu Sep 15 14:40:09 2022 daemon.notice hostapd: Configuration file: /var/run/hostapd-phy0.conf (phy wlan0) --> new PHY
Thu Sep 15 14:40:10 2022 daemon.notice hostapd: wlan0: interface state UNINITIALIZED->COUNTRY_UPDATE
Thu Sep 15 14:40:10 2022 daemon.err hostapd: WPA-PSK setup failed.
Thu Sep 15 14:40:10 2022 daemon.err hostapd: Interface initialization failed
Thu Sep 15 14:40:10 2022 daemon.notice hostapd: wlan0: interface state COUNTRY_UPDATE->DISABLED
Thu Sep 15 14:40:10 2022 daemon.notice hostapd: wlan0: AP-DISABLED
Thu Sep 15 14:40:10 2022 daemon.err hostapd: wlan0: Unable to setup interface.
Thu Sep 15 14:40:10 2022 daemon.notice hostapd: nl80211: deinit ifname=wlan0 disabled_11b_rates=0
Thu Sep 15 14:40:10 2022 daemon.err hostapd: rmdir[ctrl_interface=/var/run/hostapd]: Permission denied
Thu Sep 15 14:40:10 2022 daemon.notice hostapd: wlan0: CTRL-EVENT-TERMINATING
Thu Sep 15 14:40:10 2022 daemon.err hostapd: hostapd_free_hapd_data: Interface wlan0 wasn't started
Thu Sep 15 14:40:10 2022 daemon.notice hostapd: Configuration file: /var/run/hostapd-phy1.conf (phy wlan1) --> new PHY
Thu Sep 15 14:40:10 2022 daemon.notice hostapd: wlan1: interface state UNINITIALIZED->COUNTRY_UPDATE
Thu Sep 15 14:40:10 2022 daemon.notice hostapd: wlan1: interface state COUNTRY_UPDATE->HT_SCAN
Thu Sep 15 14:40:10 2022 daemon.notice netifd: radio0 (2463): Command failed: ubus call hostapd config_add {"iface":"wlan0", "config":"/var/run/hostapd-phy0.conf"} (Invalid argument)
Thu Sep 15 14:40:10 2022 daemon.notice hostapd: wlan1: interface state HT_SCAN->DFS
Thu Sep 15 14:40:10 2022 daemon.notice hostapd: wlan1: DFS-CAC-START freq=5560 chan=112 sec_chan=-1, width=1, seg0=106, seg1=0, cac_time=60s
Thu Sep 15 14:46:26 2022 daemon.notice hostapd: wlan1: DFS-CAC-COMPLETED success=1 freq=5560 ht_enabled=0 chan_offset=0 chan_width=3 cf1=5530 cf2=0
Thu Sep 15 14:46:26 2022 daemon.err hostapd: WPA-PSK setup failed.
Thu Sep 15 14:46:26 2022 daemon.err hostapd: Interface initialization failed
Thu Sep 15 14:46:26 2022 daemon.notice hostapd: wlan1: interface state DFS->DISABLED
Thu Sep 15 14:46:26 2022 daemon.notice hostapd: wlan1: AP-DISABLED
Thu Sep 15 14:46:28 2022 daemon.notice hostapd: wlan1: DFS-PRE-CAC-EXPIRED freq=5500 ht_enabled=0 chan_offset=0 chan_width=0 cf1=5500 cf2=0
Thu Sep 15 14:46:28 2022 daemon.notice hostapd: wlan1: DFS-PRE-CAC-EXPIRED freq=5520 ht_enabled=0 chan_offset=0 chan_width=0 cf1=5520 cf2=0
Thu Sep 15 14:46:28 2022 daemon.notice hostapd: wlan1: DFS-PRE-CAC-EXPIRED freq=5540 ht_enabled=0 chan_offset=0 chan_width=0 cf1=5540 cf2=0
Thu Sep 15 14:46:28 2022 daemon.notice hostapd: wlan1: DFS-PRE-CAC-EXPIRED freq=5560 ht_enabled=0 chan_offset=0 chan_width=0 cf1=5560 cf2=0
```

wifi is non functional.

**Comment by gstrauss** [comment]
> @gstrauss I think the proper way forward is to do testing on the malta target with hwsim(enabled by default).

I did not see malta as a target in `make menuconfig`.  Any additional hints?

I found one problem in the code:
hostap uses IKE ids (Diffie-Hellman Group Transform IDs), not TLS group ids (previously EC Named Curve Registry).
https://www.iana.org/assignments/ikev2-parameters/ikev2-parameters.xhtml
https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-8
Even though they both describe a similar set of EC curves, the values are different -- close but different.  I had to write my own mapping from IKE ids to TLS ids for mbedtls and now additional SAE tests pass for me (from the hwsim test suite).  I'll post updated patches tomorrow evening.  Are you able to regenerate the patches if I force push to my development branches, or do you prefer an increasing number of additional fixup patches?

**Comment by neheb** [comment]
Yeah. That's how I did this PR. I have your branch cloned right now.

**Comment by gstrauss** [comment]
Thanks.  I see the MIPS Malta.  Sorry, I had overlooked it.

Currently, I am using a qemu image of Debian Bullseye with hwsim (enabled by default) in order to run the hwsim tests.
Any advantage to me switching to running the hwsim tests in the MIPS malta target?  (I would have to do a complete rebuild of my openwrt environment)  ...I suppose this will eventually be necessary for me to install the openwrt test packages for mbedtls and hostap, and then run hostapd.

**Comment by neheb** [comment]
Off the top of my head advantages:

musl compatibility?
big endian (ignore the 64-bit above. it doesn't work).

That's about it.

**Comment by AuthorReflex** [comment]
I have old AP TP-Link WA901ND with small flash. I build mbedTLS variant for ath79 and run on it. It's small and works perfect on this AP without errors

**Comment by gstrauss** [comment]
I just updated and force pushed to my openwrt-hostap-vers branch with fixes for mapping IKE to mbedtls curve group.

@neheb: Please regenerate the patches and replace the existing ones.

More SAE and DPP hwsim tests pass and most OWE (opportunistic wireless encryption) hwsim tests now pass, too.
Hopefully these patches will get you up and running, though further work is likely needed for SAE and DPP.

My time to work on this will be more limited the rest of this week.
I'll try to put together a MIPS malta test environment this coming weekend.

I expect that I will need to submit a patch for openwrt mbedtls package to reduce the crippling previously wrought on the openwrt mbedtls package.  If mbedtls is to be a low level TLS and crypto *library* used by more and more openwrt applications, then mbedtls must not be crippled for the limited use only by specific applications the original cripplers had in mind when they crippled the openwrt mbedtls package.

@AuthorReflex: thanks for the data point.  We'll incrementally make it better. :)   Which hostap variant are you running?  `*-basic-mbedtls`, `*-mesh-mbedtls`, or `*-full-mbedtls`?  Please post if you find critical features missing or misbehaving.  The `*-full-mbedtls` variant of hostap, wpa_supplicant is not feature complete.  I have not yet implemented some large features that are present in the `*-full-openssl` variant but not the `*-full-mbedtls` variant.

**Comment by AuthorReflex** [comment]
@gstrauss I build && run basic variant

**Comment by gstrauss** [comment]
@neheb I plan to continue working on improving the SAE support later this week and next week.  That said, for prioritization purposes, what are the (known) requirements to get this PR out of WIP?  I think minimal requirements should be for the hostap `*-basic-mbedtls` variants to work sufficiently.  I do not believe this must include SAE, DPP, or EAP-FAST support, so we might already be there.

For `*-full-mbedtls`, SAE support is definitely a requirement.  On the other hand, is DPP or EAP-FAST support a requirement for `*-full-mbedtls`?  I do not think so, but I would like to know if others do find it important.  Again, these questions are for prioritization.

I do not currently plan to try to implement 100% of hostap features using mbedtls (though if a large, enterprise company wanted to sponsor that work, I could).  Among other things, I am not working on DPP2 or DPP3 support, though I did sketch out basic support and requirements for DPP2 near the bottom of crypto_mbedtls.c.

**Comment by ynezz** [comment]
> I think minimal requirements should be for the hostap *-basic-mbedtls variants to work sufficiently.

I agree. We need to make sure, that commit message makes it clear, what is expected to be working, what is work in progress or not supported yet.

>  I do not believe this must include SAE

FYI since 21.02.0 release there is [WPA3 available as default feature](https://openwrt.org/releases/21.02/notes-21.02.0#wpa3_support_included_by_default), so it would be probably nice to have it, but not strictly necessary.

**Comment by neheb** [comment]
I would think the next version will be in 2023. So plenty of time to figure things out.

@gstrauss IMO WPA2-PSK is good enough for that. As this PR is, it doesn't even run.

**Comment by neheb** [comment]
Rebased to latest master. Had to get rid of patch 080.

**Comment by gstrauss** [comment]
> FYI since 21.02.0 release there is [WPA3 available as default feature](https://openwrt.org/releases/21.02/notes-21.02.0#wpa3_support_included_by_default), so it would be probably nice to have it, but not strictly necessary.

Since the current Makefile for openwrt hostap package contains `DRIVER_MAKEOPTS += CONFIG_TLS=mbedtls CONFIG_SAE=y` (and similar for openssl and wolfssl), I guess we should consider SAE a requirement.

I am interested to hear if SAE is now working for @neheb with his latest push.

**Comment by neheb** [comment]
SAE is not working no.

WPA2-PSK is working. Log before and after SAE:

```
~# logread | grep hostapd
Mon Sep 19 23:30:03 2022 daemon.notice hostapd: Configuration file: /var/run/hostapd-phy0.conf (phy wlan0) --> new PHY
Mon Sep 19 23:30:04 2022 daemon.notice hostapd: wlan0: interface state UNINITIALIZED->COUNTRY_UPDATE
Mon Sep 19 23:30:04 2022 daemon.err hostapd: WPA-PSK setup failed.
Mon Sep 19 23:30:04 2022 daemon.err hostapd: Interface initialization failed
Mon Sep 19 23:30:04 2022 daemon.notice hostapd: wlan0: interface state COUNTRY_UPDATE->DISABLED
Mon Sep 19 23:30:04 2022 daemon.notice hostapd: wlan0: AP-DISABLED
Mon Sep 19 23:30:04 2022 daemon.err hostapd: wlan0: Unable to setup interface.
Mon Sep 19 23:30:04 2022 daemon.notice hostapd: nl80211: deinit ifname=wlan0 disabled_11b_rates=0
Mon Sep 19 23:30:04 2022 daemon.err hostapd: rmdir[ctrl_interface=/var/run/hostapd]: Permission denied
Mon Sep 19 23:30:04 2022 daemon.notice hostapd: wlan0: CTRL-EVENT-TERMINATING
Mon Sep 19 23:30:04 2022 daemon.err hostapd: hostapd_free_hapd_data: Interface wlan0 wasn't started
Mon Sep 19 23:30:04 2022 daemon.notice hostapd: Configuration file: /var/run/hostapd-phy1.conf (phy wlan1) --> new PHY
Mon Sep 19 23:30:04 2022 daemon.notice netifd: radio0 (2452): Command failed: ubus call hostapd config_add {"iface":"wlan0", "config":"/var/run/hostapd-phy0.conf"} (Invalid argument)
Mon Sep 19 23:30:04 2022 daemon.notice hostapd: wlan1: interface state UNINITIALIZED->COUNTRY_UPDATE
Mon Sep 19 23:30:04 2022 daemon.notice hostapd: wlan1: interface state COUNTRY_UPDATE->HT_SCAN
Mon Sep 19 23:30:04 2022 daemon.notice hostapd: wlan1: interface state HT_SCAN->DFS
Mon Sep 19 23:30:04 2022 daemon.notice hostapd: wlan1: DFS-CAC-START freq=5560 chan=112 sec_chan=-1, width=1, seg0=106, seg1=0, cac_time=60s
Tue Sep 20 00:34:45 2022 daemon.notice hostapd: wlan1: DFS-CAC-COMPLETED success=1 freq=5560 ht_enabled=0 chan_offset=0 chan_width=3 cf1=5530 cf2=0
Tue Sep 20 00:34:45 2022 daemon.err hostapd: WPA-PSK setup failed.
Tue Sep 20 00:34:45 2022 daemon.err hostapd: Interface initialization failed
Tue Sep 20 00:34:45 2022 daemon.notice hostapd: wlan1: interface state DFS->DISABLED
Tue Sep 20 00:34:45 2022 daemon.notice hostapd: wlan1: AP-DISABLED
Tue Sep 20 00:34:47 2022 daemon.notice hostapd: wlan1: DFS-PRE-CAC-EXPIRED freq=5500 ht_enabled=0 chan_offset=0 chan_width=0 cf1=5500 cf2=0
Tue Sep 20 00:34:47 2022 daemon.notice hostapd: wlan1: DFS-PRE-CAC-EXPIRED freq=5520 ht_enabled=0 chan_offset=0 chan_width=0 cf1=5520 cf2=0
Tue Sep 20 00:34:47 2022 daemon.notice hostapd: wlan1: DFS-PRE-CAC-EXPIRED freq=5540 ht_enabled=0 chan_offset=0 chan_width=0 cf1=5540 cf2=0
Tue Sep 20 00:34:47 2022 daemon.notice hostapd: wlan1: DFS-PRE-CAC-EXPIRED freq=5560 ht_enabled=0 chan_offset=0 chan_width=0 cf1=5560 cf2=0
Tue Sep 20 00:35:36 2022 daemon.notice hostapd: Remove interface 'wlan1'
Tue Sep 20 00:35:36 2022 daemon.notice hostapd: wlan1: interface state DISABLED->DISABLED
Tue Sep 20 00:35:36 2022 daemon.notice hostapd: wlan1: AP-DISABLED
Tue Sep 20 00:35:36 2022 daemon.notice hostapd: wlan1: CTRL-EVENT-TERMINATING
Tue Sep 20 00:35:36 2022 daemon.err hostapd: rmdir[ctrl_interface=/var/run/hostapd]: Permission denied
Tue Sep 20 00:35:36 2022 daemon.err hostapd: hostapd_free_hapd_data: Interface wlan1 wasn't started
Tue Sep 20 00:35:36 2022 daemon.notice hostapd: nl80211: deinit ifname=wlan1 disabled_11b_rates=0
Tue Sep 20 00:35:36 2022 daemon.notice hostapd: nl80211: Failed to remove interface wlan1 from bridge br-lan: Invalid argument
Tue Sep 20 00:35:37 2022 daemon.notice hostapd: Configuration file: /var/run/hostapd-phy0.conf (phy wlan0) --> new PHY
Tue Sep 20 00:35:37 2022 daemon.notice hostapd: wlan0: interface state UNINITIALIZED->COUNTRY_UPDATE
Tue Sep 20 00:35:37 2022 daemon.notice hostapd: wlan0: interface state COUNTRY_UPDATE->ENABLED
Tue Sep 20 00:35:37 2022 daemon.notice hostapd: wlan0: AP-ENABLED
Tue Sep 20 00:35:37 2022 daemon.notice hostapd: Configuration file: /var/run/hostapd-phy1.conf (phy wlan1) --> new PHY
Tue Sep 20 00:35:38 2022 daemon.notice hostapd: wlan1: interface state UNINITIALIZED->COUNTRY_UPDATE
Tue Sep 20 00:35:38 2022 daemon.notice hostapd: wlan1: interface state COUNTRY_UPDATE->HT_SCAN
Tue Sep 20 00:35:38 2022 daemon.notice hostapd: wlan1: interface state HT_SCAN->DFS
Tue Sep 20 00:35:38 2022 daemon.notice hostapd: wlan1: DFS-CAC-START freq=5560 chan=112 sec_chan=-1, width=1, seg0=106, seg1=0, cac_time=60s
Tue Sep 20 00:35:45 2022 daemon.info hostapd: wlan0: STA a8:e6:21:eb:b0:bb IEEE 802.11: authenticated
Tue Sep 20 00:35:45 2022 daemon.info hostapd: wlan0: STA a8:e6:21:eb:b0:bb IEEE 802.11: associated (aid 1)
Tue Sep 20 00:35:45 2022 daemon.notice hostapd: wlan0: AP-STA-CONNECTED a8:e6:21:eb:b0:bb
Tue Sep 20 00:35:45 2022 daemon.info hostapd: wlan0: STA a8:e6:21:eb:b0:bb RADIUS: starting accounting session 1E81B7904085748E
Tue Sep 20 00:35:45 2022 daemon.info hostapd: wlan0: STA a8:e6:21:eb:b0:bb WPA: pairwise key handshake completed (RSN)
Tue Sep 20 00:35:45 2022 daemon.notice hostapd: wlan0: EAPOL-4WAY-HS-COMPLETED a8:e6:21:eb:b0:bb
```

**Comment by neheb** [comment]
Yeah...my phone is connected to wifi using 11n. ax needs SAE.

**Comment by gstrauss** [comment]
> Yeah...my phone is connected to wifi using 11n. ax needs SAE.

Yes, I believe that ax needs WPA3.

While most of the hwsim sae tests are passing, most of the sae_pk hwsim tests are not, and I have not yet gotten up to testing and troubleshooting that.

**Comment by gstrauss** [comment]
This was a doozy of a bug to find: I found a single instance in hostap/src/common/sae.c where the second and fourth params to `crypto_bignum_exptmod()` matched, and that caused a problem with the underlying call to `mbedtls_mpi_exp_mod()` where the result param must not match the exponent param since the result param is used to store intermediate results.  In the current mbedtls implementation, it is ok for the the first and fourth params to `crypto_bignum_exptmod()` to match, and that is used in a number of other places.  In any case, the single use in sae.c where second and fourth params to `crypto_bignum_exptmod()` matched was used in a binary (boolean) selection, and so it did not always fail if the result bit happened to be correct.  That took a while to track down in the crypto calculations.

tl;dr: SAE should now work more reliably with what I just forced-pushed to my development branches.

103/111 hwsim sae test cases pass (with 4 skipped and 4 failing).

Note: hwsim sae_pk test cases are still failin, and I plan to look into those over the next few days.

**Comment by neheb** [comment]
@gstrauss note that your openwrt branch needs rebasing. OpenWrt master currently has hostapd commit b704dc72ef824dfdd96674b90179b274d1d38105

**Comment by gstrauss** [comment]
Thanks.  I rebased my `openwrt-hostap-vers` branch onto b704dc72ef824dfdd96674b90179b274d1d38105

**Comment by neheb** [comment]
still not working unfortunately:

```
Sun Sep 25 01:17:45 2022 daemon.notice hostapd: Configuration file: /var/run/hostapd-phy1.conf (phy wlan1) --> new PHY
Sun Sep 25 01:17:45 2022 daemon.notice hostapd: wlan1: interface state UNINITIALIZED->COUNTRY_UPDATE
Sun Sep 25 01:17:45 2022 daemon.notice hostapd: wlan1: interface state COUNTRY_UPDATE->HT_SCAN
Sun Sep 25 01:17:45 2022 daemon.notice hostapd: Configuration file: /var/run/hostapd-phy0.conf (phy wlan0) --> new PHY
Sun Sep 25 01:17:45 2022 daemon.notice hostapd: wlan0: interface state UNINITIALIZED->COUNTRY_UPDATE
Sun Sep 25 01:17:45 2022 daemon.err hostapd: WPA-PSK setup failed.
Sun Sep 25 01:17:45 2022 daemon.err hostapd: Interface initialization failed
Sun Sep 25 01:17:45 2022 daemon.notice hostapd: wlan0: interface state COUNTRY_UPDATE->DISABLED
Sun Sep 25 01:17:45 2022 daemon.notice hostapd: wlan0: AP-DISABLED
Sun Sep 25 01:17:45 2022 daemon.err hostapd: wlan0: Unable to setup interface.
Sun Sep 25 01:17:45 2022 daemon.notice hostapd: nl80211: deinit ifname=wlan0 disabled_11b_rates=0
Sun Sep 25 01:17:45 2022 daemon.err hostapd: rmdir[ctrl_interface=/var/run/hostapd]: Permission denied
Sun Sep 25 01:17:45 2022 daemon.notice hostapd: wlan0: CTRL-EVENT-TERMINATING
Sun Sep 25 01:17:45 2022 daemon.err hostapd: hostapd_free_hapd_data: Interface wlan0 wasn't started
Sun Sep 25 01:17:45 2022 daemon.notice netifd: radio0 (2460): Command failed: ubus call hostapd config_add {"iface":"wlan0", "config":"/var/run/hostapd-phy0.conf"} (Invalid argument)
Sun Sep 25 01:17:45 2022 daemon.notice hostapd: wlan1: interface state HT_SCAN->DFS
Sun Sep 25 01:17:45 2022 daemon.notice hostapd: wlan1: DFS-CAC-START freq=5560 chan=112 sec_chan=-1, width=1, seg0=106, seg1=0, cac_time=60s

```

**Comment by neheb** [comment]
Am I supposed to have https://github.com/Mbed-TLS/mbedtls/pull/6282 ?

**Comment by gstrauss** [comment]
> Am I supposed to have https://github.com/Mbed-TLS/mbedtls/pull/6282 ?

No.  The supported code is in my patches, and would need to remain part of my patches unless `#ifdef`'d away for future versions of mbedtls.

`Sun Sep 25 01:17:45 2022 daemon.err hostapd: WPA-PSK setup failed.`

It is unlikely that my patches would pass almost all of the (positive and negative) hwsim tests (running on Debian Bullseye) for sae and then have SAE not work at all on openwrt due to my patches, so if you get a chance, I would suggest testing on openwrt with an openwrt mbedtls package which as not been crippled.  Then again, I have not gotten up to testing hwsim sae_pk tests, so the WPA3 PSK might not yet be working.  TBD.

**Comment by neheb** [comment]
hostapd config file generated by netifd:

```
driver=nl80211
logger_syslog=127
logger_syslog_level=2
logger_stdout=127
logger_stdout_level=2
country_code=US
ieee80211d=1
hw_mode=g
supported_rates=60 90 120 180 240 360 480 540
basic_rates=60 120 240
beacon_int=100
dtim_period=2
channel=1
chanlist=1


ieee80211n=1
ht_coex=0
ht_capab=[LDPC][SHORT-GI-20][SHORT-GI-40][TX-STBC][RX-STBC1][MAX-AMSDU-7935]
ieee80211ax=1
he_bss_color=128
he_su_beamformer=1
he_mu_beamformer=1
he_default_pe_duration=4
he_rts_threshold=1023
he_mu_edca_qos_info_param_count=0
he_mu_edca_qos_info_q_ack=0
he_mu_edca_qos_info_queue_request=0
he_mu_edca_qos_info_txop_request=0
he_mu_edca_ac_be_aifsn=8
he_mu_edca_ac_be_aci=0
he_mu_edca_ac_be_ecwmin=9
he_mu_edca_ac_be_ecwmax=10
he_mu_edca_ac_be_timer=255
he_mu_edca_ac_bk_aifsn=15
he_mu_edca_ac_bk_aci=1
he_mu_edca_ac_bk_ecwmin=9
he_mu_edca_ac_bk_ecwmax=10
he_mu_edca_ac_bk_timer=255
he_mu_edca_ac_vi_ecwmin=5
he_mu_edca_ac_vi_ecwmax=7
he_mu_edca_ac_vi_aifsn=5
he_mu_edca_ac_vi_aci=2
he_mu_edca_ac_vi_timer=255
he_mu_edca_ac_vo_aifsn=5
he_mu_edca_ac_vo_aci=3
he_mu_edca_ac_vo_ecwmin=5
he_mu_edca_ac_vo_ecwmax=7
he_mu_edca_ac_vo_timer=255

radio_config_id=b4cfb1f9e54d66c3950bf03ef102fda5
interface=wlan0
ctrl_interface=/var/run/hostapd
ap_isolate=1
bss_load_update_period=60
chan_util_avg_period=600
disassoc_low_ack=1
skip_inactivity_poll=0
preamble=1
wmm_enabled=1
ignore_broadcast_ssid=0
uapsd_advertisement_enabled=1
utf8_ssid=1
multi_ap=0
sae_require_mfp=1
sae_pwe=2
wpa_passphrase=XXX
wpa_psk_file=/var/run/hostapd-wlan0.psk
auth_algs=1
wpa=2
wpa_pairwise=CCMP
ssid=Neheb
bridge=br-lan
wds_bridge=
snoop_iface=br-lan
wpa_disable_eapol_key_retries=0
wpa_key_mgmt=WPA-PSK WPA-PSK-SHA256 SAE
okc=1
ieee80211w=1
group_mgmt_cipher=AES-128-CMAC
dynamic_vlan=0
vlan_naming=1
vlan_no_bridge=1
vlan_file=/var/run/hostapd-wlan0.vlan
qos_map_set=0,0,2,16,1,1,255,255,18,22,24,38,40,40,44,46,48,56
config_id=b29978ffc1676829fd3fd33d96b2a3a8
bssid=XXX
```

**Comment by gstrauss** [comment]
Trying not to lose momentum, but some more urgent IRL things took priority.

New push this evening contains patches for EAP and SAE-PK.  Most tests are now passing in these areas.

(Exception: EAP-FAST and EAP-TEAP use a SessionTicket mechanism which I have not yet implemented, but those should not be required for proper SAE functioning)

Gory details: SAE-PK requires the DER containing the subject public key include the point in compressed point format, which was not well documented in the interface `crypto_ec_key_get_subject_public_key` in `crypto.h`, and `mbedtls_pk_write_pubkey_der()` uses the uncompressed point format.  I'll submit a patch upstream to update the interface comment in `crypto.h`.  It took me longer than I'd like to trace through the crypto steps to find the reason why there was a mismatch in the K_AP fingerprint.

I'll have built OpenWRT for MIPS Malta and will finish setting that up for testing later this week.

**Comment by neheb** [comment]
do note mac80211.sh seems broken on malta. so editing /etc/config/wireless will probably not be fruitful.

**Comment by neheb** [comment]
This time it's working. WPA2/3 mixed mode. I'll test actual WPA3 later today, but so far both interfaces are initialized and my phone (WPA2) is connecting.

**Comment by robimarko** [comment]
SAE is working for me on on IPQ8074(ath11k)

**Comment by neheb** [comment]
Laptop's connecting to ax. Looks like it's working here.

**Comment by neheb** [comment]
Since this is usable now, unmarking as draft.

**Comment by gstrauss** [comment]
@neheb I was looking back at conversations from 2021 and 2022 regarding code size of crypto libraries.  There are features of mbedtls which are disabled in the openwrt mbedtls package.  Similarly, there are features of wolfSSL which are disabled in the openwrt wolfSSL package.  Would it be feasible -- or would there perhaps be too much complexity or confusion -- to have a -basic and a -full variant of mbedtls, and a -basic and a -full variant of wolfssl?  The -full-mbedtls variant of hostap could depend on the -full variant of mbedtls, and the hostap -full-wolfssl variant could depend on the -full variant of wolfSSL.

I guess that openwrt package variants are for applications, and are not recommended for libraries, but I wanted to ask.  (An application depending on mbedtls-basic could also use mbedtls-full if installed on a system, but an application depending on mbedtls-full might fail to start if the mbedtls-basic package were updated over the mbedtls-full package on a system.)

---

Regardless of the above, the hostap package could probably use a review and discussion to see about disabling ancient features.  For example: should `CONFIG_EAP_MD5` be omitted in the future?  Another example is TLS1.0 and TLS1.1 support.  Forward looking, mbedtls 3.x drops support for TLS1.0 and TLS1.1.

**Comment by neheb** [comment]
wolf and openssl can be configured at build time with menucinfig. Mbedtls cannot. I assume it'd be difficult to implement.

**Comment by Neustradamus** [comment]
@neheb: A nice PR!

@gstrauss: Good job!

**Comment by gstrauss** [comment]
Status: most hwsim tests now pass.  Latest pushes round out DPP support (not DPP2 or DPP3), and add support for CRLs and certificate verification according to custom rules exposed by hostapd/wpa_supplicant configuration interfaces.

Of the 3900+ tests under tests/hwsim, fewer than 25 fail (aka: "do not pass") on my system, not including the handful that also fail using openssl on my system, and not including the tests which are skipped due to missing features (e.g. mbedtls does not yet support TLS 1.3 or OCSP, and I have not yet implemented specialized session ticket support for EAP-FAST and EAP-TEAP).  Of the tests that are failing, more than a few are due to the testing framework not getting a specific message rather than failing functionality.

Without trying to explain each test, I'll summarize and say that I am not aware of any major missing or broken functionality required for this PR to proceed.

I consider the code to be an alpha release candidate, where everything that is expected to work should work, and known limitations or features documented as not implemented are not expected to work.

While I plan to continue polishing as time allows, I think the next step is to encourage wider testing and gather feedback if there are bugs or required features which are missing.  I'll try again to get some traction on the hostap mailing list.

@neheb: please update the patches in the PR when you have a moment.  Thanks!

**Comment by neheb** [comment]
Updated.

Anyway, router's uptime is 8 days. I don't see any issues with mixed WPA2/WPA3 mode.

**Comment by gstrauss** [comment]
Latest push; updating patches is not urgent.  Added stricter enforcement of SUITEB192, added detection of certificate policies (TOD-STRICT, TOD-TOFU), fixed IKEv2 (algorithm selection mismatch), enabled some additional features in the test suite and made adjustments so that tests pass.

**Comment by gstrauss** [comment]
Re-posted to hostap mailing list: http://lists.infradead.org/pipermail/hostap/2022-October/040891.html

**Comment by gstrauss** [comment]
@neheb wrote
> wolf and openssl can be configured at build time with menucinfig. Mbedtls cannot. I assume it'd be difficult to implement.

Not difficult, but would take some effort.  Modifying the Makefile for menuconfig options and using mbedtls `scripts/config.py [set|unset] ...` should probably be a separate PR to replace `openwrt/package/libs/mbedtls/patches/200-config.patch`
Run `scripts/config.py --help` in the mbedtls source tree for more info.  I think that is a bit out of scope for this PR.

**Comment by gstrauss** [comment]
For future consideration:
`CONFIG_FILS=y`
`CONFIG_FILS_SK_PFS=y`
might be enabled for variant=full.

This was previously discussed in 2019:
https://forum.openwrt.org/t/how-to-enable-fils-fast-initial-link-setup-defined-in-the-ieee80211ai-standard-feature-in-the-openwrt-wi-fi-ap-mode/42327/2
The landscape has improved since then, and FILS (Fast Initial Link Setup) is probably a good idea in dense locations with multiple access points, which probably overlaps with places where OWE (opportunistic wireless encryption) is used.  (`CONFIG_OWE=y` is currently enabled for hostapd basic and full variants.)

More details about FILS can be found at
https://go.7signal.com/80211ai-fast-initial-link-setup-whitepaper

**Comment by gstrauss** [comment]
Status: PR #11033 filed to improve mbedtls build options through menuconfig.  If PR #11033 is accepted, that will remove the patch in this PR for mbedtls to enable DES.  Then again, with my current development branch, you do not have to modify mbedtls to enable DES since hostap provides an internal implementation of `des_encrypt()`, and I modified by development branch to use that hostap internal implementation if mbedtls was built without DES support.

**Comment by neheb** [comment]
well, added the commit to this PR for now.

**Comment by neheb** [comment]
Updated to latest git progress.

**Comment by aparcar** [comment]
Could you please rebase this?

**Comment by neheb** [comment]
Applied latest patchset. Although there hasn't been an update in 24 days.

**Comment by gstrauss** [comment]
FYI: there are no important fixes on my local dev branch, though I have done some polishing for future-proofing with mbedtls 3.x (by reducing use of `MBEDTLS_PRIVATE`).  These changes have not yet been pushed publicly.

Discussion on hostap mailing lists suggests that there is a limitation/incompatibility with DPP and the current implementation requirements of `mbedtls_ecp_mul()`.  Reminder: DPP is not a critical requirement for the hostap mbedtls variant.  On the other hand, parsing of PKCS#7 has been committed to mbedtls 3.x development branch, so my code for hostap mbedtls now theoretically supports DPP2 with a future mbedtls 3.3.0 (?) release, except for the aforementioned limitation/incompatibility wth `mbedtls_ecp_mul()`.

---
@neheb what do you think of my question further up about
`CONFIG_FILS=y`
`CONFIG_FILS_SK_PFS=y`
?

**Comment by neheb** [comment]
Sounds good to enable. I assume out of scope for this PR.

**Comment by gstrauss** [comment]
> Sounds good to enable. I assume out of scope for this PR.

FILS is not required for this PR.

However, the code supports FILS and the hwsim tests pass.  The modification in this PR would be to enable FILS in the Makefile for hostap-full-mbedtls, and perhaps also for hostap-basic-mbedtls.

Out-of-scope for this PR: enabling and testing FILS for openssl and wolfssl.

**Comment by neheb** [comment]
Ehhh probably would make more sense to do a PR for the others first and then mbedtls.

**Comment by gstrauss** [comment]
In 2019: FILS was not available in OpenWRT due to `package/kernel/mac80211/patches/subsys/130-disable-fils.patch`
How to enable FILS (Fast Initial Link setup, defined in the IEEE80211ai standard) feature in the OpenWrt Wi-Fi AP mode?
https://forum.openwrt.org/t/how-to-enable-fils-fast-initial-link-setup-defined-in-the-ieee80211ai-standard-feature-in-the-openwrt-wi-fi-ap-mode/42327

In 2021: 53b6783907f3bd6f0f88f9d6feed20b21e2cd181, Felix removed patches which disabled FILS support in the kernel, making FILS available.

@nbd168: in 5b66dfaf6c53ac21285cbabd8935666d34d6378a, you enabled `CONFIG_FILS=y` in files/wpa_supplicant-p2p.config and files/wpa_supplicant-full.config, but not files/hostapd-full.config.  Do you remember why?  Also, why not also enable `CONFIG_FILS_SK_PFS=y`?

**Comment by dangowrt** [comment]
I've been test-running this on some of my local APs and it works great. (Tested: WPA2 PSK and WPA3 SAE, both STA and AP mode, all works as expected). Is there anything which prevents us from merging this now?
I agree with @neheb that FILS and FILS_SK_PFS can be enabled for all relevant variants in a follow-up PR.

**Comment by neheb** [comment]
@dangowrt I don't see an issue with merging as is.

**Comment by dangowrt** [comment]
Pulled to master https://github.com/openwrt/openwrt/commit/6d1df35747d7f02f52973ba5604642dcc17ddf5f

**Comment by gstrauss** [comment]
In the interest of potentially getting additional testing and use before considering changing hostapd mbedtls variant to be the default in a future openwrt release, would you consider backporting these patches to openwrt 22.03 to make the hostapd mbedtls variant an option there (*not the default* in 22.03)?

**Comment by Neustradamus** [comment]
Thanks @neheb and @dangowrt!

It is possible to do a backport?
